File: services/api/payments/provider.py
"""Contrat PaymentProvider — playbook §10.2. Ne pas modifier sans ADR."""

from __future__ import annotations

from typing import Literal, Protocol


class PaymentOrder(dict):
    """tenant_slug, amount_xof, channel, idem_key, customer_country=BJ."""


class PaymentProvider(Protocol):
    id: Literal["geniuspay", "demo"]

    def create_transaction(self, order: PaymentOrder) -> dict: ...
    def verify(self, reference: str) -> dict | None: ...
    def handle_webhook(self, raw_body: str, signature: str, timestamp: str) -> dict | None: ...
    def refund(self, reference: str, amount_xof: int, reason: str) -> dict: ...
    def reconcile(self, date_from: str, date_to: str) -> dict: ...

File: apps/web/src/lib/pay/provider.ts
import type { LedgerEntry, PaymentChannel } from "@/lib/demo/store";

/** Contrat maison — playbook §10.2. Le métier ne parle jamais au SDK Genius Pay. */
export type PaymentOrder = {
  tenantSlug: string;
  amountXof: number;
  channel: PaymentChannel;
  idemKey: string;
  customerCountry?: "BJ";
};

export type WebhookHeaders = {
  signature: string;
  timestamp: string;
  event?: string;
  environment?: string;
};

export type RefundResult = {
  ok: boolean;
  reference: string;
  error?: string;
};

export type ReconcileRow = {
  reference: string;
  amountXof: number;
  matched: boolean;
};

export type ReconcileResult = {
  from: string;
  to: string;
  rows: ReconcileRow[];
  discrepancies: string[];
};

export interface PaymentProvider {
  readonly id: "geniuspay" | "demo" | "api";
  createTransaction(order: PaymentOrder): Promise<LedgerEntry>;
  verify(reference: string): Promise<LedgerEntry | undefined>;
  handleWebhook(
    rawBody: string,
    headers: WebhookHeaders,
  ): Promise<LedgerEntry | undefined>;
  refund(
    reference: string,
    amountXof: number,
    reason: string,
  ): Promise<RefundResult>;
  reconcile(range: { from: string; to: string }): Promise<ReconcileResult>;
}
